slashing-protection/src/main/resources/migrations/postgresql/V00012__add_highwatermark_metadata.sql

```diff
@@ -0,0 +1,63 @@
+ALTER TABLE metadata
+    ADD COLUMN high_watermark_epoch NUMERIC(20),
+    ADD COLUMN high_watermark_slot NUMERIC(20);
+
+-- inserted high watermark should be above low watermark
+
+CREATE OR REPLACE FUNCTION check_high_watermarks() RETURNS TRIGGER AS $$
+DECLARE
+    max_slot NUMERIC(20);
+    max_epoch NUMERIC(20);
+BEGIN
+SELECT MAX(slot) INTO max_slot FROM low_watermarks;
+SELECT GREATEST(MAX(target_epoch), MAX(source_epoch)) INTO max_epoch FROM low_watermarks;
+
+IF NEW.high_watermark_slot <= max_slot THEN
+        RAISE EXCEPTION 'Insert/Update violates constraint: high_watermark_slot must be greater than max slot in low_watermarks table';
+END IF;
+
+IF NEW.high_watermark_epoch <= max_epoch THEN
+        RAISE EXCEPTION 'Insert/Update violates constraint: high_watermark_epoch must be greater than max epoch in low_watermarks table';
+END IF;
+
+RETURN NEW;
+END;
+$$ LANGUAGE plpgsql;
+
+CREATE TRIGGER check_before_insert_or_update_high_watermarks
```

Comment:
You are using a trigger because you want to enforce a constraint across the metadata and low_watermark tables?

Reply:
It's not possible to use CHECK CONSTRAINT across tables with postgres unfortunately: https://www.postgresql.org/docs/current/ddl-constraints.html (the note at the bottom of section 5.4.1)